implement README parsing

SaketramDurbha opened the pull request:
This PR implements functionality to parse terminal commands for building and deploying a sample to the Cloud Run fully managed platform out of the sample's README.md. Code blocks that include commands that the serverless-sample-tester would like to run should be immediately preceded with a comment including the tag `sst-run-linuxmacos`. For example:

````text
[//]: # ({sst-run-linuxmacos})
```bash
gcloud builds submit --tag=gcr.io/${GOOGLE_CLOUD_PROJECT}/run-mysql
```
````

This PR does **not** include:
-  Revised algorithm to replace Cloud Run service names in parsed terminal commands.
-  Customizable Cloud Run region and revised hierarchy for customization of Cloud Run region.

These changes will be made in (a) separate PR(s).

Commit:
do not set region flag for gcloud run; rely on gcloud run/region property

Commit:
add README parsing

Commit:
minor updates to README parsing documentation and logging

Commit:
minor updates to README parsing code

Commit:
remove extraneous empty line

Commit:
fix code style in internal/lifecycle/readme.go

Commit:
add extra instructions to README.md

Commit:
add wrapper error message to error in lifecycle.parseREADME

Commit:
fix typo in README.md

Commit:
update README.md readme parsing example

jsimonweb commented on README.md:
```diff
@@ -25,4 +25,31 @@ go build -o sst cmd/main.go
+Do not set the Cloud Run region you'd like to deploy to through the `--region` flag in the `gcloud run` commands.
+Instead, as mentioned above, do so by setting the `run/region` gcloud property.
+
+In addition to setting a default Cloud Run region, make sure to deploy to the fully platform on Cloud Run. You can
```
' ... fully platform ' 

should be

' ... fully managed platform '

SaketramDurbha replied:
Resolved, thank you.

jsimonweb commented on README.md:
```diff
@@ -25,4 +25,31 @@ go build -o sst cmd/main.go
+achieve this by setting the `run/platform` gcloud property to `managed` or passing in the `--platform=managed` flag
+to your `gcloud run` commands.
+
+If code tags aren't added to your README, the program will fall back to reasonable defaults to build and deploy your
```
"If code tags ..."

update to be

"If comment code tags ..."

SaketramDurbha replied:
Resolved, thank you.

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,127 @@
+
+// The tag that should appear immediately before code blocks in a README to indicate that the enclosed commands
+// are to be used by this program for building and deploying the sample.
+var codeTag = "sst-run-linuxmacos"
```
Does this tag need to include `linuxmacos`? Shouldn't gcloud commands be the same for either Windows or Unix?

SaketramDurbha replied:
I think it would be the same for Unix. But I wasn't not sure how the `exec.Command` and associated go interfaces.

One thing that concerns me specifically is that os.ExpandEnv seems to only expand variables with the `$VAR` or `${VAR}` syntax. Whereas it appears that Windows Command Prompt users can expand variables using the `%VAR%` syntax.

dmahugh replied:
Might want to just focus on bash syntax for now, since Windows syntax has a couple of variations. Windows CMD prompt expands environment variables with ```%varname%``` syntax, and PowerShell (which is more commonly used these days) expands them with ```$env:varname``` syntax.

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,127 @@
+
+	inCode := false
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
```
Do we need to Trim every line? That's a lot of work when we are only using the changed line some of the time

SaketramDurbha replied:
Resolved, thank you.

kurtisvg commented on internal/lifecycle/readme.go:
````diff
@@ -0,0 +1,127 @@
+		line := strings.TrimSpace(scanner.Text())
+
+		if strings.Contains(line, "```") && inCode {
+			inCode = false
````
Do we care about when we are inCode unless we are already in a codeTag?

kurtisvg commented on README.md:
````diff
@@ -25,4 +25,31 @@ go build -o sst cmd/main.go
+the program should build and deploy your sample:
+
+```text
+[//]: # ({sst-run-linuxmacos})
````
What's the significance of `[//] #`?

SaketramDurbha replied:
This, in my understanding, is one syntax for a line comment in Markdown. It's the same syntax used here: https://raw.githubusercontent.com/googleapis/java-bigtable/master/README.md , for a similar purpose. 

However, formal documentation for Markdown comments has been hard to find. But it appears there are multiple syntaxes for line comments.

dmahugh replied:
This syntax is widely supported and probably the right choice here, but comment syntax in Markdown isn't standardized, unfortunately. FYI, this SO thread has a lot of info: https://stackoverflow.com/questions/4823468/comments-in-markdown

kurtisvg replied:
SG, but probably worth mentioning that in some fashion in the README

kurtisvg replied:
To clarify, we should clarify what the user needs to do - is it a commented line? which kind of comment do we support? Can it be added somewhere else? We want to minimize the risk of a false positive with this.

SaketramDurbha replied:
Okay, should we require code tags follow this regular expression pattern?:

```
\[//\]:\s+#\s+\(\s*{sst-run-linuxmacos}\s*\)
```

This would require a comment in the form of the `[//] # (comment)` syntax. We could also require that a code tag line following this format _immediately_ precede a code block.

kurtisvg replied:
To clarify, I think given the ambiguity of the comment syntax I agree with your initial decision to use a looser requirement. However, you should make sure that what you are looking for is clearly defined in the README. So something like:

> When parsing the README for custom build/deploy instructions, SST will include any commands inside a code fence that is immediately preceded by a line containing `{sst-run-unix}`. You can use a markdown comment syntax (e.g. `[//] # ({sst-run-unix})`) to include this line without making it visible when rendered.

This makes it clear that:
1. The line only has to contain `{sst-run-unix}`
2. There can't be any empty lines between the line containing the tag and the code block.
3. They can use `[//] #`, but the exact format is up to them. 

You may also which to clarify the following points:
* Are multiline commands supported?
* Is bash syntax supported (`cd` or `export VAR=VAL`)?

SaketramDurbha replied:
README updated, thank you.

kurtisvg commented on internal/lifecycle/readme.go:
````diff
@@ -0,0 +1,127 @@
+
+		if strings.Contains(line, "```") && inCode {
+			inCode = false
+		} else if strings.Contains(line, codeTag) {
````
Is `Contains` all we want? Shouldn't the line be pretty much exactly `[//] # SST_RUN_xxx` once trimmed?

SaketramDurbha replied:
As mentioned above (https://github.com/GoogleCloudPlatform/serverless-sample-tester/pull/2#discussion_r465461157), it appears there are multiple syntaxes for line comments in Mardown. That's why the liberal option of just looking for a mention the specific code tag in the line is used. I'm not sure whether we should just require this line specifically.

And if we stick with the initial implementation, after some thought, I'm now thinking that we shouldn't thrown an error if we aren't able to find the start of a code block on the next line; we should just ignore the current line with a code tag.

SaketramDurbha replied:
Also, this, I believe, is the same way that the releasetool looks for package version update tags/markers in READMEs like the ones found here:
https://raw.githubusercontent.com/googleapis/java-bigtable/master/README.md

releasetool source: https://github.com/googleapis/releasetool/blob/master/releasetool/commands/start/java.py#L229

kurtisvg replied:
Looks like the source you've linked matches `{<some_tag>}` (not just `<some_tag>`). I think it would make sense to be more explicit (again to minimize false positives)

SaketramDurbha replied:
Sorry, my mistake. I've update the `codeTag` constant to be `{sst-run-unix}`

kurtisvg commented on internal/lifecycle/readme.go:
````diff
@@ -0,0 +1,127 @@
+			startCodeBlockLine := scanner.Text()
+
+			if !strings.Contains(startCodeBlockLine, "```") {
+				return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: incorrect format")
````
Errors should be informative and ideally actionable. This say "incorrect format" - how does the user act to fix that?

SaketramDurbha replied:
Error messages updated, thank you.

kurtisvg commented on internal/lifecycle/readme.go:
````diff
@@ -0,0 +1,127 @@
+	for scanner.Scan() {
+		line := strings.TrimSpace(scanner.Text())
+
+		if strings.Contains(line, "```") && inCode {
````
You should check `inCode` first - `Contains` is a more expensive operation

SaketramDurbha replied:
Irrelevant after refactor based on @tritone review.

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,127 @@
+
+			inCode = true
+		} else if inCode {
+			for line[len(line)-1] == '\\' {
```
A comment would be good here - I'm not sure I understand what's happening here

kurtisvg replied:
Is this to support multi line code?

SaketramDurbha replied:
Yes -- added comment to clarify this, thank you.

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,127 @@
+// It detects whether the command is a gcloud run command and replaces the last argument that isn't a flag
+// with the input service name.
+func replaceServiceName(command, serviceName string) string {
+	containsGcloud, _ := regexp.MatchString(`\bgcloud\b`, command)
```
Probably worth a comment on why we ignore these errors

SaketramDurbha replied:
Will add a comment. Also an explanation here:

The only scenario in which the `regexp.MatchString` function will return an error is if the regex fails to compile. But in our case, since we're using the same two pieces of regex each time and we know that they're valid, it seemed like we could safely ignore these errors.

`regexp.MatchString` source: https://github.com/golang/go/blob/master/src/regexp/regexp.go#L548 .

Given this, I don't know if it's acceptable to be ignoring the errors or if we should be catching the errors for good measure anyways.

kurtisvg replied:
A better alternative here would be to save the regex as a constant, so it's calculated at compile time. You can use `regex.MustCompile()` and then `(*Regexp).MatchString`. This will help you avoid the case where someone modifies the regex and you don't find out it's invalid until your compiler runs.

SaketramDurbha replied:
Sorry, do you mean something like this?

```
r1 := regexp.MustCompile(`\bgcloud\b`)
r2 := regexp.MustCompile(`\brun\b`)

containsGcloud:= r1.MatchString(command)
containsRun := r2.MatchString(command)
```

kurtisvg replied:
Yup!

SaketramDurbha replied:
Resolved, thank you.

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,127 @@
+	sp := strings.Split(command, " ")
+	for i := len(sp) - 1; i >= 0; i-- {
+		if !strings.Contains(sp[i], "--") {
+			sp[i] = serviceName
```
Can gcloud flags have args or are they always in the `flag=val` format?

```
gcloud run --flag1 arg1 serviceName --flag2 arg2
```

SaketramDurbha replied:
I believe they can have args and can be in the format followed in your example.

We (@danielylee01) were to revise the Cloud Run service name detection and replacing algorithm in a future PR so as to separate the work each of us do into separate PRs.

kurtisvg replied:
SGTM - let's open an issue then for this PR to make sure we don't lose track. We can close it with a future PR once complete.

SaketramDurbha replied:
Okay, planning to open an issue entitled "Cloud Run service name detection and replacing algorithm flawed" after this current PR gets merged.

kurtisvg commented on internal/lifecycle/readme.go:
````diff
@@ -0,0 +1,127 @@
+		if strings.Contains(line, "```") && inCode {
+			inCode = false
+		} else if strings.Contains(line, codeTag) {
+			scanner.Scan()
````
What if this is false?

SaketramDurbha replied:
Updated to check for return value of all `Scanner.Scan()` calls and check `Scanner.Err()` if appropriate.

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,127 @@
+	var lifecycle Lifecycle
+
+	inCode := false
+	for scanner.Scan() {
```
You need to check `scanner.Err` I think?

SaketramDurbha replied:
I think the `bufio.Scanner` `Scan` method returns a bool indicating whether there was an error or at the end of file. If the bool indicates there is potentially an error, then we can check `Scanner.Err` to see what that error message specifically is.

Documentation here: https://golang.org/pkg/bufio/#Scanner.Scan .

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,127 @@
+		return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: no commands found")
+	}
+
+	return lifecycle, nil
```
nit: This whole function might be more readable if you allow for some nesting. Some example psuedo:
```
for line in scanner {
     // check scanner.Err
     line := scanner.Text()
     if line contains code_tag {
          for command in code_block {
                // parse command
                // convert command
           }
     }
}
```

SaketramDurbha replied:
Mostly irrelevant after refactor based on @tritone review, thank you.

dmahugh reviewed (approved):
LGTM (modulo comments from others), added some info on two topics.

Commit:
fix typo and add clarification to README.md

Commit:
add error message to error in lifecycle/readme.go

Commit:
address reviews for lifecycle/readme.go

Commit:
change code tag from sst-run-linuxmacos -> sst-run-unix

Commit:
update codeTag in code to more closely match instructions in README.md

SaketramDurbha commented:
Per a recommendation from @kurtisvg , we've updated the code tag from `sst-run-linuxmacos` to `sst-run-unix`.

Commit:
refactor regexp matching code in lifecycle/readme.go

Commit:
add clarification for REAMDE parsing in README.md

- clarity on expected syntax/location of code tags
- clarify commands aren't run through a shell
- add info on mutliline commands and env var expansion

kurtisvg commented on internal/lifecycle/lifecycle.go:
```diff
@@ -40,22 +41,39 @@ func (l Lifecycle) Execute(commandsDir string) error {
+	_, err := os.Stat(readmePath)
+	readmeExists := err == nil
+
+	if readmeExists {
```
nit: I think you can remove this variable from scope altogether now 
```suggestion
	

	if _, err := os.Stat(readmePath); err != nil {
```

SaketramDurbha replied:
Resolved, thank you.

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,149 @@
+				} else { // scanner.Scan falied, but no error found. EOF.
+					return nil, fmt.Errorf("[lifecycle.parseREADME] parsing README: unexpected EOF; file ended immediately after code tag")
+				}
+			}
```
Can this be simplified? It looks like :
```
if !s or !c {
    if err and !s -> scanner err
    if c -> not code block err
    else -> EoF err
}
```

I think this is simplified as: 

```
if !s {
   if err -> scanner err
   else -> EoF err
}
if !c {
   -> not code block err
}
```

SaketramDurbha replied:
Resolved, thank you.

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,149 @@
+				for line[len(line)-1] == '\\' {
+					line = line[:len(line)-1]
+
+					scanner.Scan()
```
What happens if this Scan is false? (or contains "```"?)

SaketramDurbha replied:
Same as here: https://github.com/GoogleCloudPlatform/serverless-sample-tester/pull/2#discussion_r467312902 . Thank you.

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,149 @@
+// It detects whether the command is a gcloud run command and replaces the last argument that isn't a flag
+// with the input service name.
+func replaceServiceName(command, serviceName string) string {
+	r1 := regexp.MustCompile(`\bgcloud\b`)
+	r2 := regexp.MustCompile(`\brun\b`)
```
nit: You should move these to global vars (so they are constants, and not re initialized each time).

also: better variables names if global constants

SaketramDurbha replied:
Resolved, thank you.

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,149 @@
+	containsGcloud := r1.MatchString(command)
+	containsRun := r2.MatchString(command)
+
+	if !(containsGcloud && containsRun) {
```
nit: You can avoid adding these variables to scope as well
```suggestion
	if !(containsGcloud.MatchString(command) && containsRun.MatchString(command)) {
```

SaketramDurbha replied:
Resolved, thank you.

Commit:
address reviews part two

tritone commented on internal/lifecycle/lifecycle.go:
```diff
@@ -40,9 +41,21 @@ func (l Lifecycle) Execute(commandsDir string) error {
+// those options are set up, it falls back to reasonable defaults based on whether the sample is java-based
 // (has a pom.xml) that doesn't have a Dockerfile or isn't.
 func NewLifecycle(sampleDir, serviceName, gcrURL string) Lifecycle {
+	readmePath := fmt.Sprintf("%s/README.md", sampleDir)
```
You can use path/filepath for path manipulations (here and below). See https://gobyexample.com/file-paths and https://golang.org/pkg/path/filepath/

SaketramDurbha replied:
Resolved, thank you.

tritone commented on internal/lifecycle/lifecycle.go:
```diff
@@ -40,9 +41,21 @@ func (l Lifecycle) Execute(commandsDir string) error {
+			log.Println("Using build and deploy commands found in README")
+			return lifecycle
+		}
+		fmt.Printf("Failed parsing README: %v\n", err)
```
I would think it would be preferable to have a fatal error here if the README exists but is invalid, rather than continuing to the backup methods.

SaketramDurbha replied:
Resolved. Now we're only continuing to backup methods when either (a) the README.md file doesn't exist in the root directory or (b) it's obvious that the user hasn't tried to set the README up to work with our program (there are no tagged code blocks). This is per @kurtisvg suggestion in chat.

tritone commented on internal/lifecycle/lifecycle.go:
```diff
@@ -53,9 +66,11 @@ func NewLifecycle(sampleDir, serviceName, gcrURL string) Lifecycle {
 	dockerfileExists := err == nil
```
Technically there can be other errors from os.Stat other than the file not existing. It's probably fine not to distinguish but the variable names are slightly misleading.

SaketramDurbha replied:
Resolved, thank you.

tritone commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,156 @@
+
+// parseREADME parses a README file with the given name. It reads terminal commands surrounded by one of the codeTags
+// listed above and loads them into a Lifecycle. In the process, it replaces the Cloud Run service name and Container
+// Registry tag with the provided inputs.
```
General comment looking at this function: parsing logic is pretty complicated and can be very tricky to get right in all the edge cases. How well-defined is the README format? Is doing this parsing a hard requirement for your project? If possible it would simplify things a lot if you could force users to provide input data as a json and then rely on the nice built in support that golang has for parsing those.

SaketramDurbha replied:
Per offline discussion, we've decided that we'll keep pursuing it, but make sure to spend more time thinking about potential edge cases and keep an eye out for any that may come up. Also, will open an issue after this PR is merged for tests that will help us also help us catch edge cases.

tritone commented on internal/lifecycle/readme.go:
````diff
@@ -0,0 +1,156 @@
+			}
+
+			startCodeBlockLine := scanner.Text()
+			c := strings.Contains(startCodeBlockLine, "```")
````
I'd also use a constant for the triple backticks (and any other magic values like this).

SaketramDurbha replied:
Resolved, thank you.

tritone commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,156 @@
+var (
+	// The tag that should appear immediately before code blocks in a README to indicate that the enclosed commands
+	// are to be used by this program for building and deploying the sample.
+	codeTag = "{sst-run-unix}"
```
Seems like these should be const.

SaketramDurbha replied:
Resolved, thank you.

tritone commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,156 @@
+				}
+
+				line = os.ExpandEnv(line)
+				line = gcrURLRegexp.ReplaceAllString(line, gcrURL)
```
Is it possible to separate out the logic involved in just reading the readme from this text substitution? It would involve an additional pass over the data but I think it would make the code clearer (and it doesn't seem like efficiency is the biggest concern here). In this case you could do something like:

* Scan the file and generate a slice of strings that correspond to the code blocks.
* Then write a separate function that turns a code block string into an exec.Cmd by doing substitutions, dealing with multi-line commands, etc

You could even have a separate data type for a code block (`type codeBlock string`) and then give it a method that returns the command.

tritone replied:
Also I think this could have additional benefits as far as making testing easier, and allowing for support of other input formats other than README should that become necessary.

SaketramDurbha replied:
Resolved, thank you.

Commit:
refactor code based on code review

Commit:
update error messages and error checking in README

only use reasonable defaults if README doesn't exist or code tags aren't included in README

Commit:
refactor README parsing code according to code review

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,205 @@
+
+		if strings.Contains(line, codeTag) {
+			if s := scanner.Scan(); !s {
+				if err := scanner.Err(); err != nil && !s {
```
nit: I think `!s` is already true if we're in this section?
```suggestion
				if err := scanner.Err(); err != nil {
```

SaketramDurbha replied:
Resolved, thank you.

Commit:
update error handling for README parsing

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,205 @@
+				if err := scanner.Err(); err != nil && !s {
+					return nil, fmt.Errorf("[lifecycle.codeBlocks] bufio.Scanner: %w", err)
+				}
+				return nil, fmt.Errorf("[lifecycle.codeBlocks]: unexpected EOF; file ended immediately " +
+					"after code tag")
```
nit: Go _generally_ doesn't enforce a line limit, so it's typically fine to let a long line happen

If you really prefer to break the line, it's better to do something like this:
```go
return nil, fmt.Errorf(
       "[lifecycle.codeBlocks]: unexpected EOF; file ended immediately after code tag")
```
or this:
```go
s := fmt.Errorf("[lifecycle.codeBlocks]: unexpected EOF; file ended immediately after code tag")
return nil, s
```

The reason for this is that breaking in the middle of the string makes it harder to read - you're leaving it up to the user to mentally parse out the `" + "` and make sure everything lines up correctly.

SaketramDurbha replied:
Resolved, thank you

kurtisvg commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,206 @@
+			sp := strings.Split(line, " ")
+
+			var cmd *exec.Cmd
+			if strings.Contains(line, "gcloud") {
```
Should this be "contains" or `sp[0] == "gcloud"`?

SaketramDurbha replied:
Resolved, thank you

Commit:
fix typo in lifecycle.go

Commit:
address reviews

- remove line breaks for long strings
- other small changes

tritone commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,201 @@
+		return nil, errNoREADMECodeBlocksFound
+	}
+
+	lifecycle, err := codeBlocksTolifecycle(codeBlocks, serviceName, gcrURL)
```
I would define a custom type for codeBlock as a slice of strings to make it more clear what's going on. The surface would then look something like this (names might need a bit of thought...):

```
type codeBlock []string

func (cb codeBlock)  ToCommand(serviceName, gcrURL string) (*exec.Cmd, error) {...}

func extractBlocks(scanner *bufio.Scanner) ([]codeBlock, error) {...}
```

Then in parseREADME the logic would go (very basic sketch here):

```
blocks, err := extractBlocks(scanner)

// handle err

var l Lifecycle
for _, b := range blocks {
   c, err := b.toCommand(serviceName, gcrURL)
   // handle err
  l = append(l, c)
}

```

This allows you to re-use the code block parsing if there is some other source of input in the future, make it easier to test, and would better clarify what is happening where.

SaketramDurbha replied:
Thank you, resolved.

tritone commented on internal/lifecycle/readme.go:
```diff
@@ -0,0 +1,201 @@
+
+				i++
+				if i >= len(block) {
+					return nil, fmt.Errorf("[lifecycle.codeBlocksTolifecycle]: unexpected end of code block; expecting command line continuation")
```
no need for brackets around the function name in the error msg.

SaketramDurbha replied:
Thanks, planning to do this in a separate PR because other error messages outside of this PR are also formatted like this.

SaketramDurbha replied:
Opening an issue for this after PR is merged.

tritone commented on internal/sample/sample.go:
```diff
@@ -66,7 +66,10 @@ func NewSample(dir string) (*Sample, error) {
-	buildDeployLifecycle := lifecycle.NewLifecycle(dir, service.Name, cloudContainerImageURL)
+	buildDeployLifecycle, err := lifecycle.NewLifecycle(dir, service.Name, cloudContainerImageURL)
+	if err != nil {
+		return nil, fmt.Errorf("[sample.NewSample] getting Lifecycle: %w", err)
```
Again, no need for brackets

SaketramDurbha replied:
Same as here: https://github.com/GoogleCloudPlatform/serverless-sample-tester/pull/2#discussion_r467152252 . Thank you.

tritone reviewed (commented):
Nice improvement. I would also recommend adding tests to this PR if possible-- it will make it clearer what the right surface design is, and help you think about edge cases for your parsing logic.

Commit:
refactor README parsing to use new codeBlock struct
- create codeBlock struct and receiver method
- modify lifecycle readme functions to use codeBlock struct

tritone commented on internal/lifecycle/readme.go:
```diff
@@ -65,19 +118,24 @@ func parseREADME(filename, serviceName, gcrURL string) (Lifecycle, error) {
-		return lifecycle, fmt.Errorf("[lifecycle.parseREADME] transforming code blocks in %s to executable commands: %w", filename, err)
+	var l Lifecycle
+	for _, b := range codeBlocks {
+		c, err := b.toCommands(serviceName, gcrURL)
```
Maybe rename `c` to something plural to indicate that it is a slice rather than a single command.

SaketramDurbha replied:
Resolved, thank you.

tritone commented on internal/lifecycle/readme.go:
```diff
@@ -65,19 +118,24 @@ func parseREADME(filename, serviceName, gcrURL string) (Lifecycle, error) {
+			return l, fmt.Errorf("[lifecycle.parseREADME] transforming code blocks in %s to executable commands: %w", filename, err)
+		}
+
+		l = append(l, c...)
```
Looking at this logic, I think that properly speaking you should redefine `NewLifecycle` to take a slice of codeBlock as a means of initializing, so you always have the same entry point. It breaks the encapsulation somewhat to have a NewLifecycle function but then have other ways of constructing a Lifecycle as well. The surface would be something like:

```
// Surfaces:
type LifecycleOption interface {
  apply(l *Lifecycle) // this method applies the option to the lifecycle.
}

type withCodeBlocks struct {
  cb []CodeBlock
}

func (w withCodeBlocks) apply(l *Lifecycle) {
  // this does the logic that you have in L122-129. 
}

func WithCodeBlocks(blocks []CodeBlock) LifecycleOption {
  // returns a withCodeBlocks struct using the given blocks.
}

// additional functions to produce appropriate lifecycle options as needed.

func NewLifecycle(opts ...LifecycleOption)  // This could have additional params such as service name if needed.

// To use:
l := NewLifecycle(FromCodeBlocks(blocks))
```

I don't think you have to fix this in order to merge this PR, but this is the kind of pattern that I would expect to see in order to polish this up.

tritone replied:
Discussed this offline with Saketram-- I totally got the logic backwards here (since actually NewLifecycle is what calls the readme parsing logic), sorry about that!

I think it would be preferable for NewLifecycle to have options specifying what kind of lifecycle is desired rather than using defaults based on the environment or what files are present, but I'm not sure if that's fully compatible with project requirements.

Commit:
rename var name to more closely align with var's function in readme.go
